### ISSUE TITLE
Cannot filter on joined/linked table when question has Custom Column

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#12129]
**Describe the bug**
When creating a question with a Custom Column, it's not possible to filter on linked/joined tables.

**To Reproduce**
Linked table:
1. Notebook question > Sample Dataset > Orders
2. Create Custom Column `[Discount] * 2` called "Mega Discount"
3. Filter by Product.Category=Doohickey
4. Errors with `Column "PRODUCTS__via__PRODUCT_ID.CATEGORY" not found`

Joined table:
1. Notebook question > Sample Dataset > Orders
2. Join table Products on Orders.ProductID=Products.ID
3. Create Custom Column `[Discount] * 2` called "Mega Discount"
4. Filter by Product.Category=Doohickey
5. Errors with `Column "Product.CATEGORY" not found;`

**Information about your Metabase Installation:**
`master` on several platforms with different datasources - it used to work on 0.34.3

**Comment by tlrobinson** [comment]
@sbelak @camsaul AFAICT this is a regression in the query processor.

The `json_query` looks right to me

```json
  {
    "type": "query",
    "query": {
      "source-table": 2,
      "expressions": { "Mega Discount": ["*", ["field-id", 10], 2] },
      "filter": ["=", ["fk->", ["field-id", 11], ["field-id", 6]], "Doohickey"],
      "limit": 10
    },
    "database": 1,
    "parameters": [ ],
    "middleware": { "add-default-userland-constraints?": true }
  }
```

but not the SQL:

```sql
SELECT
    "source"."ID" AS "ID",
    "source"."CREATED_AT" AS "CREATED_AT",
    "source"."DISCOUNT" AS "DISCOUNT",
    "source"."PRODUCT_ID" AS "PRODUCT_ID",
    "source"."QUANTITY" AS "QUANTITY",
    "source"."SUBTOTAL" AS "SUBTOTAL",
    "source"."TAX" AS "TAX",
    "source"."TOTAL" AS "TOTAL",
    "source"."USER_ID" AS "USER_ID",
    "source"."Mega Discount" AS "Mega Discount",
    "source"."TITLE" AS "TITLE",
    "source"."NAME" AS "NAME"
FROM (
        SELECT
            ("PUBLIC"."ORDERS"."DISCOUNT" * 2) AS "Mega Discount",
            "PUBLIC"."ORDERS"."DISCOUNT" AS "DISCOUNT",
            "PRODUCTS__via__PRODUCT_ID"."CATEGORY" AS "CATEGORY",
            "PUBLIC"."ORDERS"."ID" AS "ID",
            "PUBLIC"."ORDERS"."CREATED_AT" AS "CREATED_AT",
            "PUBLIC"."ORDERS"."PRODUCT_ID" AS "PRODUCT_ID",
            "PUBLIC"."ORDERS"."QUANTITY" AS "QUANTITY",
            "PUBLIC"."ORDERS"."SUBTOTAL" AS "SUBTOTAL",
            "PUBLIC"."ORDERS"."TAX" AS "TAX",
            "PUBLIC"."ORDERS"."TOTAL" AS "TOTAL",
            "PUBLIC"."ORDERS"."USER_ID" AS "USER_ID",
            "PRODUCTS__via__PRODUCT_ID"."TITLE" AS "TITLE",
            "PEOPLE__via__USER_ID"."NAME" AS "NAME",
            "PRODUCTS__via__PRODUCT_ID"."ID" AS "ID_2",
            "PEOPLE__via__USER_ID"."ID" AS "ID_3"
        FROM "PUBLIC"."ORDERS"
        LEFT JOIN "PUBLIC"."PRODUCTS" "PRODUCTS__via__PRODUCT_ID" ON "PUBLIC"."ORDERS"."PRODUCT_ID" = "PRODUCTS__via__PRODUCT_ID"."ID"
        LEFT JOIN "PUBLIC"."PEOPLE" "PEOPLE__via__USER_ID" ON "PUBLIC"."ORDERS"."USER_ID" = "PEOPLE__via__USER_ID"."ID"
    ) "source"
WHERE "PRODUCTS__via__PRODUCT_ID"."CATEGORY" = ? LIMIT 1
```

Full dataset request

```json
{
  "database_id": 1,
  "started_at": "2020-03-18T11:12:14.721561-07:00",
  "state": "42S22",
  "json_query": {
    "type": "query",
    "query": {
      "source-table": 2,
      "expressions": { "Mega Discount": ["*", ["field-id", 10], 2] },
      "filter": ["=", ["fk->", ["field-id", 11], ["field-id", 6]], "Doohickey"],
      "limit": 10
    },
    "database": 1,
    "parameters": [ ],
    "middleware": { "add-default-userland-constraints?": true }
  },
  "native": {
    "query": "SELECT \"source\".\"ID\" AS \"ID\", \"source\".\"CREATED_AT\" AS \"CREATED_AT\", \"source\".\"DISCOUNT\" AS \"DISCOUNT\", \"source\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"source\".\"QUANTITY\" AS \"QUANTITY\", \"source\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"source\".\"TAX\" AS \"TAX\", \"source\".\"TOTAL\" AS \"TOTAL\", \"source\".\"USER_ID\" AS \"USER_ID\", \"source\".\"Mega Discount\" AS \"Mega Discount\", \"source\".\"TITLE\" AS \"TITLE\", \"source\".\"NAME\" AS \"NAME\" FROM (SELECT (\"PUBLIC\".\"ORDERS\".\"DISCOUNT\" * 2) AS \"Mega Discount\", \"PUBLIC\".\"ORDERS\".\"DISCOUNT\" AS \"DISCOUNT\", \"PRODUCTS__via__PRODUCT_ID\".\"CATEGORY\" AS \"CATEGORY\", \"PUBLIC\".\"ORDERS\".\"ID\" AS \"ID\", \"PUBLIC\".\"ORDERS\".\"CREATED_AT\" AS \"CREATED_AT\", \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"PUBLIC\".\"ORDERS\".\"QUANTITY\" AS \"QUANTITY\", \"PUBLIC\".\"ORDERS\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"PUBLIC\".\"ORDERS\".\"TAX\" AS \"TAX\", \"PUBLIC\".\"ORDERS\".\"TOTAL\" AS \"TOTAL\", \"PUBLIC\".\"ORDERS\".\"USER_ID\" AS \"USER_ID\", \"PRODUCTS__via__PRODUCT_ID\".\"TITLE\" AS \"TITLE\", \"PEOPLE__via__USER_ID\".\"NAME\" AS \"NAME\", \"PRODUCTS__via__PRODUCT_ID\".\"ID\" AS \"ID_2\", \"PEOPLE__via__USER_ID\".\"ID\" AS \"ID_3\" FROM \"PUBLIC\".\"ORDERS\" LEFT JOIN \"PUBLIC\".\"PRODUCTS\" \"PRODUCTS__via__PRODUCT_ID\" ON \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" = \"PRODUCTS__via__PRODUCT_ID\".\"ID\" LEFT JOIN \"PUBLIC\".\"PEOPLE\" \"PEOPLE__via__USER_ID\" ON \"PUBLIC\".\"ORDERS\".\"USER_ID\" = \"PEOPLE__via__USER_ID\".\"ID\") \"source\" WHERE \"PRODUCTS__via__PRODUCT_ID\".\"CATEGORY\" = ? LIMIT 10",
    "params": ["Doohickey"]
  },
  "status": "failed",
  "class": "class org.h2.jdbc.JdbcSQLException",
  "stacktrace": [
    "org.h2.message.DbException.getJdbcSQLException(DbException.java:357)",
    "org.h2.message.DbException.get(DbException.java:179)",
    "org.h2.message.DbException.get(DbException.java:155)",
    "org.h2.expression.ExpressionColumn.optimize(ExpressionColumn.java:150)",
    "org.h2.expression.Comparison.optimize(Comparison.java:177)",
    "org.h2.command.dml.Select.prepare(Select.java:861)",
    "org.h2.command.Parser.prepareCommand(Parser.java:283)",
    "org.h2.engine.Session.prepareLocal(Session.java:611)",
    "org.h2.engine.Session.prepareCommand(Session.java:549)",
    "org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1247)",
    "org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:76)",
    "org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:1135)",
    "com.mchange.v2.c3p0.impl.NewProxyConnection.prepareStatement(NewProxyConnection.java:267)",
    "--> driver.sql_jdbc.execute$eval80418$fn__80419.invoke(execute.clj:239)",
    "driver.sql_jdbc.execute$prepared_statement_STAR_.invokeStatic(execute.clj:257)",
    "driver.sql_jdbc.execute$prepared_statement_STAR_.invoke(execute.clj:254)",
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:387)",
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)",
    "driver.sql_jdbc$eval83147$fn__83148.invoke(sql_jdbc.clj:50)",
    "driver.h2$eval83467$fn__83468.invoke(h2.clj:80)",
    "query_processor.context$executef.invokeStatic(context.clj:59)",
    "query_processor.context$executef.invoke(context.clj:48)",
    "query_processor.context.default$default_runf.invokeStatic(default.clj:69)",
    "query_processor.context.default$default_runf.invoke(default.clj:67)",
    "query_processor.context$runf.invokeStatic(context.clj:45)",
    "query_processor.context$runf.invoke(context.clj:39)",
    "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)",
    "query_processor.reducible$pivot.invoke(reducible.clj:31)",
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__50510.invoke(mbql_to_native.clj:26)",
    "query_processor.middleware.check_features$check_features$fn__49744.invoke(check_features.clj:42)",
    "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__50709.invoke(optimize_datetime_filters.clj:133)",
    "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__55089.invoke(wrap_value_literals.clj:137)",
    "query_processor.middleware.annotate$add_column_info$fn__47684.invoke(annotate.clj:556)",
    "query_processor.middleware.permissions$check_query_permissions$fn__49603.invoke(permissions.clj:64)",
    "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__51246.invoke(pre_alias_aggregations.clj:40)",
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__49833.invoke(cumulative_aggregations.clj:61)",
    "query_processor.middleware.resolve_joins$resolve_joins$fn__51820.invoke(resolve_joins.clj:183)",
    "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__45616.invoke(add_implicit_joins.clj:245)",
    "query_processor.middleware.limit$limit$fn__50488.invoke(limit.clj:38)",
    "query_processor.middleware.format_rows$format_rows$fn__50461.invoke(format_rows.clj:81)",
    "query_processor.middleware.desugar$desugar$fn__49905.invoke(desugar.clj:22)",
    "query_processor.middleware.binning$update_binning_strategy$fn__48244.invoke(binning.clj:229)",
    "query_processor.middleware.resolve_fields$resolve_fields$fn__49397.invoke(resolve_fields.clj:24)",
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__45128.invoke(add_dimension_projections.clj:272)",
    "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__45364.invoke(add_implicit_clauses.clj:147)",
    "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__45859.invoke(add_source_metadata.clj:105)",
    "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__51471.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)",
    "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__47877.invoke(auto_bucket_datetimes.clj:125)",
    "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__49452.invoke(resolve_source_table.clj:46)",
    "query_processor.middleware.parameters$substitute_parameters$fn__51222.invoke(parameters.clj:97)",
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__49506.invoke(resolve_referenced.clj:80)",
    "query_processor.middleware.expand_macros$expand_macros$fn__50169.invoke(expand_macros.clj:158)",
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__45906.invoke(add_timezone_info.clj:15)",
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__54921.invoke(splice_params_in_response.clj:32)",
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__51488$fn__51492.invoke(resolve_database_and_driver.clj:33)",
    "driver$do_with_driver.invokeStatic(driver.clj:61)",
    "driver$do_with_driver.invoke(driver.clj:57)",
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__51488.invoke(resolve_database_and_driver.clj:27)",
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__50383.invoke(fetch_source_query.clj:243)",
    "query_processor.middleware.store$initialize_store$fn__54938$fn__54939.invoke(store.clj:11)",
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)",
    "query_processor.store$do_with_store.invoke(store.clj:40)",
    "query_processor.middleware.store$initialize_store$fn__54938.invoke(store.clj:10)",
    "query_processor.middleware.cache$maybe_return_cached_results$fn__48825.invoke(cache.clj:201)",
    "query_processor.middleware.validate$validate_query$fn__54955.invoke(validate.clj:10)",
    "query_processor.middleware.normalize_query$normalize$fn__50531.invoke(normalize_query.clj:22)",
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__45722.invoke(add_rows_truncated.clj:36)",
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__54900.invoke(results_metadata.clj:128)",
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__49768.invoke(constraints.clj:42)",
    "query_processor.middleware.process_userland_query$process_userland_query$fn__51337.invoke(process_userland_query.clj:136)",
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__49697.invoke(catch_exceptions.clj:166)",
    "query_processor.reducible$async_qp$qp_STAR___30008$thunk__30009.invoke(reducible.clj:100)",
    "query_processor.reducible$async_qp$qp_STAR___30008.invoke(reducible.clj:106)",
    "query_processor$base_qp$fn__55098.doInvoke(query_processor.clj:107)",
    "query_processor.reducible$sync_qp$qp_STAR___30017$fn__30020.invoke(reducible.clj:132)",
    "query_processor.reducible$sync_qp$qp_STAR___30017.invoke(reducible.clj:131)",
    "query_processor$process_userland_query.invokeStatic(query_processor.clj:218)",
    "query_processor$process_userland_query.doInvoke(query_processor.clj:214)",
    "query_processor$eval55159$process_query_and_save_execution_BANG___55168$fn__55171.invoke(query_processor.clj:233)",
    "query_processor$eval55159$process_query_and_save_execution_BANG___55168.invoke(query_processor.clj:225)",
    "query_processor$eval55203$process_query_and_save_with_max_results_constraints_BANG___55212$fn__55215.invoke(query_processor.clj:245)",
    "query_processor$eval55203$process_query_and_save_with_max_results_constraints_BANG___55212.invoke(query_processor.clj:238)",
    "api.dataset$fn__55241$fn__55244.invoke(dataset.clj:52)",
    "query_processor.streaming$streaming_response_STAR_$fn__41594$fn__41595.invoke(streaming.clj:73)",
    "query_processor.streaming$streaming_response_STAR_$fn__41594.invoke(streaming.clj:72)",
    "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:62)",
    "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:60)",
    "async.streaming_response$do_f_async$fn__30311.invoke(streaming_response.clj:79)"
  ],
  "context": "ad-hoc",
  "error": "Column \"PRODUCTS__via__PRODUCT_ID.CATEGORY\" not found; SQL statement:\n-- Metabase:: userID: 1 queryType: MBQL queryHash: b3894215f6b238e843f7916a5d0d8cdd2923f98ede4b5cf99937fbf262fcbb6b\nSELECT \"source\".\"ID\" AS \"ID\", \"source\".\"CREATED_AT\" AS \"CREATED_AT\", \"source\".\"DISCOUNT\" AS \"DISCOUNT\", \"source\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"source\".\"QUANTITY\" AS \"QUANTITY\", \"source\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"source\".\"TAX\" AS \"TAX\", \"source\".\"TOTAL\" AS \"TOTAL\", \"source\".\"USER_ID\" AS \"USER_ID\", \"source\".\"Mega Discount\" AS \"Mega Discount\", \"source\".\"TITLE\" AS \"TITLE\", \"source\".\"NAME\" AS \"NAME\" FROM (SELECT (\"PUBLIC\".\"ORDERS\".\"DISCOUNT\" * 2) AS \"Mega Discount\", \"PUBLIC\".\"ORDERS\".\"DISCOUNT\" AS \"DISCOUNT\", \"PRODUCTS__via__PRODUCT_ID\".\"CATEGORY\" AS \"CATEGORY\", \"PUBLIC\".\"ORDERS\".\"ID\" AS \"ID\", \"PUBLIC\".\"ORDERS\".\"CREATED_AT\" AS \"CREATED_AT\", \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"PUBLIC\".\"ORDERS\".\"QUANTITY\" AS \"QUANTITY\", \"PUBLIC\".\"ORDERS\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"PUBLIC\".\"ORDERS\".\"TAX\" AS \"TAX\", \"PUBLIC\".\"ORDERS\".\"TOTAL\" AS \"TOTAL\", \"PUBLIC\".\"ORDERS\".\"USER_ID\" AS \"USER_ID\", \"PRODUCTS__via__PRODUCT_ID\".\"TITLE\" AS \"TITLE\", \"PEOPLE__via__USER_ID\".\"NAME\" AS \"NAME\", \"PRODUCTS__via__PRODUCT_ID\".\"ID\" AS \"ID_2\", \"PEOPLE__via__USER_ID\".\"ID\" AS \"ID_3\" FROM \"PUBLIC\".\"ORDERS\" LEFT JOIN \"PUBLIC\".\"PRODUCTS\" \"PRODUCTS__via__PRODUCT_ID\" ON \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" = \"PRODUCTS__via__PRODUCT_ID\".\"ID\" LEFT JOIN \"PUBLIC\".\"PEOPLE\" \"PEOPLE__via__USER_ID\" ON \"PUBLIC\".\"ORDERS\".\"USER_ID\" = \"PEOPLE__via__USER_ID\".\"ID\") \"source\" WHERE \"PRODUCTS__via__PRODUCT_ID\".\"CATEGORY\" = ? LIMIT 10 [42122-197]",
  "row_count": 0,
  "running_time": 0,
  "preprocessed": {
    "type": "query",
    "query": {
      "source-table": 2,
      "expressions": { "Mega Discount": ["*", ["field-id", 10], 2] },
      "filter": [
        "=",
        ["joined-field", "PRODUCTS__via__PRODUCT_ID", ["field-id", 6]],
        [
          "value",
          "Doohickey",
          { "base_type": "type/Text", "special_type": "type/Category", "database_type": "VARCHAR" }
        ]
      ],
      "limit": 10,
      "fields": [
        ["field-id", 12],
        ["datetime-field", ["field-id", 15], "default"],
        ["field-id", 10],
        ["field-id", 11],
        ["field-id", 14],
        ["field-id", 13],
        ["field-id", 16],
        ["field-id", 17],
        ["field-id", 9],
        ["expression", "Mega Discount"],
        ["joined-field", "PRODUCTS__via__PRODUCT_ID", ["field-id", 5]],
        ["joined-field", "PEOPLE__via__USER_ID", ["field-id", 20]]
      ],
      "joins": [
        {
          "strategy": "left-join",
          "source-table": 1,
          "alias": "PRODUCTS__via__PRODUCT_ID",
          "fk-field-id": 11,
          "condition": ["=", ["field-id", 11], ["joined-field", "PRODUCTS__via__PRODUCT_ID", ["field-id", 4]]]
        },
        {
          "strategy": "left-join",
          "source-table": 3,
          "alias": "PEOPLE__via__USER_ID",
          "fk-field-id": 9,
          "condition": ["=", ["field-id", 9], ["joined-field", "PEOPLE__via__USER_ID", ["field-id", 21]]]
        }
      ]
    },
    "database": 1,
    "middleware": { "add-default-userland-constraints?": true },
    "info": { "executed-by": 1, "context": "ad-hoc", "nested?": false, "query-hash": "0xB3894215" },
    "constraints": { "max-results": 10000, "max-results-bare-rows": 2000 }
  },
  "data": { "rows": [ ], "cols": [ ] }
}
```

**Comment by sbelak** [comment]
Fixed via #12161

**Comment by mazameli** [comment]
@sbelak This is a little pedantic, but I don't think we should be closing issues until their corresponding fix is actually _merged_.
